doc/specs/stdlib_selection.md

```diff
@@ -0,0 +1,339 @@
+---
+title: Selection Procedures
+---
+
+# The `stdlib_selection` module
+
+[TOC]
+
+## Overview of selection
+
+Suppose you wish to find the value of the k-th smallest entry in an array of size N, or
+the index of that value. While it could be done by sorting the whole array
+using `[[stdlib_sorting(module):sort(interface)]]` or 
+`[[stdlib_sorting(module):sort_index(interface)]]` from 
+`[[stdlib_sorting(module)]]` and then finding the k-th entry, that would
+require O(N x LOG(N)) time. However selection of a single entry can be done in
+O(N) time, which is much faster for large arrays.  This is useful, for example,
+to quickly find the median of an array, or some other percentile.
+
+The Fortran Standard Library therefore provides a module, `stdlib_selection`,
+which implements selection algorithms.
+
+## Overview of the module
+
+The module `stdlib_selection` defines two generic subroutines:
+* `select` is used to find the k-th smallest entry of an array. The input
+array is also modified in-place, and on return will be partially sorted
+such that `all(array(1:k) <= array(k)))`  and `all(array(k) <= array((k+1):size(array)))`.
+The user can optionally specify `left` and `right` indices to constrain the search
+for the k-th smallest value. This can be useful if you have previously called `select`
+to find a smaller or larger rank (that will have led to partial sorting of
+`array`, thus implying some constraints on the location).
+
+* `arg_select` is used to find the index of the k-th smallest entry of an array.
+In this case the input array is not modified, but the user must provide an
+input index array with the same size as `array`, having indices that are a permutation of
+`1:size(array)`, which is modified instead. On return the index array is modified
+such that `all(array(index(1:k)) <= array(index(k)))` and `all(array(k) <= array(k+1:size(array)))`.
+The user can optionally specify `left` and `right` indices to constrain the search
+for the k-th smallest value. This can be useful if you have previously called `arg_select`
+to find a smaller or larger rank (that will have led to partial sorting of
+`index`, thus implying some constraints on the location).
+
+
+## `select` - find the k-th smallest value in an input array
+
+### Status
+
+Experimental
+
+### Description
+
+Returns the k-th smallest value of `array(:)`, and also partially sorts `array(:)`
+such that `all(array(1:k) <= array(k))`  and `all(array(k) <= array((k+1):size(array)))`
```

Comment:
A line mentioning if the sort is stable or not would be helpful.

Reply:
To my understanding the concept of stable sort does not apply here? Since we are sorting based on the value of `array` itself, not via a comparison function or some other indirect method. 

In other words, if two values of `array` are equal, they really are indistinguishable (in all contexts).

Reply:
But the point does hold for `arg_select` -- and I am not immediately sure whether it is stable, will have to look into it.

Reply:
From experiments, `arg_select` does not produce a stable sort. I have mentioned that in the documentation.